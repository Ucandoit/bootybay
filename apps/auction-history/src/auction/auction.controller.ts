import { Body, Controller, Post, UsePipes } from '@nestjs/common';
import { JoiValidationPipe } from '../commons/pipes/JoiValidationPipe';
import { AuctionService } from './auction.service';
import { createAuctionSchema } from './auction.validator';
import { CreateAuctionDto } from './dto/auction.dto';

@Controller()
export class AuctionController {
  constructor(private readonly auctionsService: AuctionService) {}

  @Post()
  @UsePipes(new JoiValidationPipe(createAuctionSchema))
  createAuction(@Body() createAuctionDto: CreateAuctionDto) {
    this.auctionsService.createAuction(createAuctionDto);
  }
}
