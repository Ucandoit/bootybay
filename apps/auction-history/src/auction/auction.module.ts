import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import {
  HistoricalAuction,
  HistoricalAuctionController,
  HistoricalAuctionRepository,
  HistoricalAuctionSchema,
  HistoricalAuctionService,
} from './historical';
import {
  RealtimeAuction,
  RealtimeAuctionController,
  RealtimeAuctionRepository,
  RealtimeAuctionSchema,
  RealtimeAuctionService,
} from './realtime';

@Module({
  imports: [
    MongooseModule.forFeature([
      { name: HistoricalAuction.name, schema: HistoricalAuctionSchema },
      { name: RealtimeAuction.name, schema: RealtimeAuctionSchema },
    ]),
  ],
  controllers: [HistoricalAuctionController, RealtimeAuctionController],
  providers: [HistoricalAuctionService, RealtimeAuctionService, HistoricalAuctionRepository, RealtimeAuctionRepository],
})
export class AuctionModule {}
