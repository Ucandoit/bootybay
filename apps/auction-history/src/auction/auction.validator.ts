import * as Joi from 'joi';

const objectSchema = Joi.object({
  itemString: Joi.string().required(),
  realm: Joi.string().required(),
  timestamp: Joi.number().required(),
  regionMarketValue: Joi.number(),
  regionHistorical: Joi.number(),
  regionSale: Joi.number(),
  regionSoldPerDay: Joi.number(),
  regionSalePercent: Joi.number(),
  marketValue: Joi.number(),
  minBuyout: Joi.number(),
  historical: Joi.number(),
  numAuctions: Joi.number(),
});

export const createAuctionSchema: Joi.AnySchema = Joi.alternatives().try(objectSchema, Joi.array().items(objectSchema));
