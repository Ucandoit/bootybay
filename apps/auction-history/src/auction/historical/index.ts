export { HistoricalAuctionController } from './historical-auction.controller';
export { HistoricalAuctionRepository } from './historical-auction.repository';
export { HistoricalAuction, HistoricalAuctionSchema } from './historical-auction.schema';
export { HistoricalAuctionService } from './historical-auction.service';
