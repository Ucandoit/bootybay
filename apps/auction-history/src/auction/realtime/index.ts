export { RealtimeAuctionController } from './realtime-auction.controller';
export { RealtimeAuctionRepository } from './realtime-auction.repository';
export { RealtimeAuction, RealtimeAuctionSchema } from './realtime-auction.schema';
export { RealtimeAuctionService } from './realtime-auction.service';
