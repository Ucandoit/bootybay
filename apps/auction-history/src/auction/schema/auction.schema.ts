import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Document } from 'mongoose';
export type AuctionDocument = Auction & Document;

@Schema({ collection: 'auctions' })
export class Auction {
  @Prop({ required: true })
  _id: string;

  @Prop({ required: true })
  itemId: number;

  @Prop({ required: true })
  realm: string;

  @Prop({ required: true })
  timestamp: number;

  @Prop()
  marketValue: number;

  @Prop()
  minBuyout: number;

  @Prop()
  historical: number;

  @Prop()
  numAuctions: number;

  @Prop()
  regionMarketValue: number;

  @Prop()
  regionHistorical: number;

  @Prop()
  regionSale: number;

  @Prop()
  regionSoldPerDay: number;

  @Prop()
  regionSalePercent: number;
}
export const AuctionSchema = SchemaFactory.createForClass(Auction);
