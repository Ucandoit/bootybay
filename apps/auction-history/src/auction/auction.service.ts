import { Injectable } from '@nestjs/common';
import { AuctionRepository } from './auction.repository';
import { CreateAuctionDto, RealmAuction, RegionalAuction } from './dto/auction.dto';
import { RecentAuctionsRequestDto, RecentAuctionsResponseDto } from './dto/recent-auctions.dto';

@Injectable()
export class AuctionService {
  constructor(private readonly auctionRepository: AuctionRepository) {}

  async createAuction(createAuctionDto: CreateAuctionDto): Promise<void> {
    if (Array.isArray(createAuctionDto)) {
      await this.auctionRepository.createMany(
        createAuctionDto.map((dto: RegionalAuction | RealmAuction) => ({
          ...dto,
          _id: `${dto.itemString}-${dto.realm}-${dto.timestamp}`,
        }))
      );
    } else {
      await this.auctionRepository.create({
        ...createAuctionDto,
        _id: `${createAuctionDto.itemString}-${createAuctionDto.realm}-${createAuctionDto.timestamp}`,
      });
    }
  }

  async getAuctionByRealm({
    realm,
    page = 0,
    size = 10,
  }: RecentAuctionsRequestDto): Promise<RecentAuctionsResponseDto> {
    const total = await this.auctionRepository.countMostRecentByRealm(realm);
    const auctions = await this.auctionRepository.findMostRecentByRealm(realm, page, size);
    return {
      page,
      size,
      total,
      auctions,
    };
  }
}
