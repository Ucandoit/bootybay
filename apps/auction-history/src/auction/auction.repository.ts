import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { AnyKeys, Model } from 'mongoose';
import { Auction, AuctionDocument } from './schema/auction.schema';

@Injectable()
export class AuctionRepository {
  constructor(@InjectModel(Auction.name) private auctionModel: Model<AuctionDocument>) {}
  async create(createEntityData: AnyKeys<unknown>): Promise<AuctionDocument> {
    return this.auctionModel.create(createEntityData);
  }

  async createMany(createEntityDatas: Array<AnyKeys<unknown>>): Promise<AuctionDocument[]> {
    return this.auctionModel.insertMany(createEntityDatas);
  }
}
