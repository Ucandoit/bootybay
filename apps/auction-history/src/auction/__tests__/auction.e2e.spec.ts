import { INestApplication } from '@nestjs/common';
import { getConnectionToken, getModelToken, MongooseModule } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import { MongoMemoryServer } from 'mongodb-memory-server';
import { Connection, Model } from 'mongoose';
import * as request from 'supertest';
import { AuctionModule } from '../auction.module';
import { Auction, AuctionDocument } from '../schema/auction.schema';
import { auctionsStub, auctionStub, realAuctionsStub, realAuctionStub } from './stubs/auction.stub';

describe('Auction e2e tests', () => {
  let app: INestApplication;
  let memoryMongodb: MongoMemoryServer;
  let model: Model<AuctionDocument>;
  let connection: Connection;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        AuctionModule,
        MongooseModule.forRootAsync({
          useFactory: async () => {
            memoryMongodb = await MongoMemoryServer.create();
            const mongoUri = await memoryMongodb.getUri();
            return {
              uri: mongoUri,
            };
          },
        }),
      ],
    }).compile();

    app = moduleRef.createNestApplication();
    connection = moduleRef.get(getConnectionToken());
    model = moduleRef.get<Model<AuctionDocument>>(getModelToken(Auction.name));
    await app.init();
  });

  beforeEach(async () => {
    await model.deleteMany({});
  });

  describe('POST /', () => {
    describe('request with invalid request body', () => {
      const testCases = [
        {
          description: 'should return 400 if body is empty',
          body: {},
          status: 400,
        },
        {
          description: 'should return 400 if required itemId is missing',
          body: {
            realm: 'bcc-eu-regional',
            timestamp: 1627580614000,
          },
          status: 400,
        },
        {
          description: 'should return 400 if required realm is missing',
          body: {
            itemId: 1,
            timestamp: 1627580614000,
          },
          status: 400,
        },
        {
          description: 'should return 400 if required timestamp is missing',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
          },
          status: 400,
        },
        {
          description: 'should return 400 if itemId is not a number',
          body: {
            itemId: 'test',
            realm: 'bcc-eu-regional',
            timestamp: 1627580614000,
          },
          status: 400,
        },
        {
          description: 'should return 400 if realm is not a string',
          body: {
            itemId: 1,
            realm: 123,
            timestamp: 1627580614000,
          },
          status: 400,
        },
        {
          description: 'should return 400 if timestamp is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if regionMarketValue is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            regionMarketValue: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if regionHistorical is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            regionHistorical: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if regionSale is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            regionSale: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if regionSoldPerDay is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            regionSoldPerDay: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if regionSalePercent is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            regionSalePercent: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if marketValue is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            marketValue: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if minBuyout is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            minBuyout: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if historical is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            historical: 'test',
          },
          status: 400,
        },
        {
          description: 'should return 400 if numAuctions is not a number',
          body: {
            itemId: 1,
            realm: 'bcc-eu-regional',
            timestamp: 'test',
            numAuctions: 'test',
          },
          status: 400,
        },
      ];

      testCases.forEach(({ description, body, status }) => {
        test(description, async () => {
          const res = await request(app.getHttpServer()).post('/').send(body).expect(status);
          expect(res.body).toMatchSnapshot();
        });
      });
    });
  });

  describe('request with valid request body', () => {
    test('should return 201 with regional auction', async () => {
      await request(app.getHttpServer()).post('/').send(auctionStub()).expect(201);
      const documents = await model.find({});
      expect(documents).toHaveLength(1);
      expect(documents[0]).toMatchObject(auctionStub());
    });

    test('should return 201 with realm auction', async () => {
      await request(app.getHttpServer()).post('/').send(realAuctionStub()).expect(201);
      const documents = await model.find({});
      expect(documents).toHaveLength(1);
      expect(documents[0]).toMatchObject(realAuctionStub());
    });

    test('should return 201 with regional auction array', async () => {
      await request(app.getHttpServer()).post('/').send(auctionsStub()).expect(201);
      const documents = await model.find({});
      expect(documents).toHaveLength(2);
      expect(documents).toMatchObject(auctionsStub());
    });

    test('should return 201 with realm auction array', async () => {
      await request(app.getHttpServer()).post('/').send(realAuctionsStub()).expect(201);
      const documents = await model.find({});
      expect(documents).toHaveLength(2);
      expect(documents).toMatchObject(realAuctionsStub());
    });

    test('unknown properties should be ignored', async () => {
      await request(app.getHttpServer())
        .post('/')
        .send({
          ...auctionStub(),
          unknown: 'unknown',
        })
        .expect(201);
      const documents = await model.find({});
      expect(documents).toHaveLength(1);
      expect(documents[0]).toMatchObject(auctionStub());
      expect(Reflect.hasOwnMetadata('unknown', documents[0])).toBe(false);
    });
  });

  afterAll(async () => {
    await connection.close();
    await memoryMongodb.stop();
    await app.close();
  });
});
