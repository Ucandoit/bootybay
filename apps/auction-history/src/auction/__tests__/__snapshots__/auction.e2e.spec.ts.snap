// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auction e2e tests POST / request with invalid request body should return 400 if body is empty 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if historical is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if itemId is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"itemId\\" must be a number",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if marketValue is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if minBuyout is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if numAuctions is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if realm is not a string 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"realm\\" must be a string",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if regionHistorical is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if regionMarketValue is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if regionSale is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if regionSalePercent is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if regionSoldPerDay is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"value\\" does not match any of the allowed types",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if required itemId is missing 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"itemId\\" is required",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if required realm is missing 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"realm\\" is required",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if required timestamp is missing 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"timestamp\\" is required",
  "statusCode": 400,
}
`;

exports[`Auction e2e tests POST / request with invalid request body should return 400 if timestamp is not a number 1`] = `
Object {
  "error": "Bad Request",
  "message": "\\"timestamp\\" must be a number",
  "statusCode": 400,
}
`;
